mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,21 +137,37 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
+      val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
+
+      val relativeError = 0.05
+      val (summary, count) = modelDatasetWithDist.select(distCol)
+        .rdd
+        .mapPartitions { iter =>
+          if (iter.hasNext) {
+            var s = new QuantileSummaries(
+              QuantileSummaries.defaultCompressThreshold, relativeError)
+            var c = 0L
+            while (iter.hasNext) {
+              val Row(dist: Double) = iter.next
+              s = s.insert(dist)
+              c += 1
+            }
+            Iterator.single((s.compress, c))
+          } else Iterator.empty
+        }.treeReduce { case ((s1, c1), (s2, c2)) => (s1.merge(s2), c1 + c2) }
 
       // Compute threshold to get around k elements.
       // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
       // so we pick quantile p = M / N + err
       // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
       val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
-      val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
+
       if (approxQuantile >= 1) {
         modelDatasetWithDist
       } else {
-        val hashThreshold = modelDatasetWithDist.stat
-          .approxQuantile(distCol, Array(approxQuantile), relativeError)
+        val hashThreshold = summary.query(approxQuantile).get
```

Comment:
`modelDatasetWithDist.stat.approxQuantile` always ignores nan values:
multipleApproxQuantiles:
```
    def apply(summaries: Array[QuantileSummaries], row: Row): Array[QuantileSummaries] = {
      var i = 0
      while (i < summaries.length) {
        if (!row.isNullAt(i)) {
          val v = row.getDouble(i)
          if (!v.isNaN) summaries(i) = summaries(i).insert(v)
        }
        i += 1
      }
      summaries
    }
```


here I should had filtered nan values. I send a followup https://github.com/apache/spark/pull/29778 for this.